src/v/cluster/log_eviction_stm.cc

```diff
@@ -97,6 +86,29 @@ ss::future<> log_eviction_stm::write_raft_snapshots_in_background() {
                   _raft->ntp());
             }
         }
+        /// This background fiber can be woken-up via apply() when special
+        /// batches are processed or by the storage layer when local
+        /// eviction is triggered.
+        try {
+            const auto eviction_event_wait_time = 5s;
+            if (index_lb && _raft->last_snapshot_index() < index_lb) {
+                /// If after call to \ref do_write_snapshot() above has not
+                /// truncated up until the desired point, it will be OK to
+                /// iterate again after some time to retry. Maybe
+                /// max_collectible_offset has moved forward.
```

Comment:
For this to matter, we need to evaluate the max collectible offset in this loop, instead of in the storage notification handler.

Reply:
index_lb will be trimmed to max collectible. If the raft snapshot is not index_lb its known we only truncated to a lower max collectible.

Reply:
Ah thanks, indeed. I think we may want to early return after resetting `index_lb` in `do_write_snapshot()`? We might be double-writing the snapshot if the max collectible offset hasn't moved

Reply:
technically that cannot happen because within `consensus::write_snapshot` there is a bounds check, but then we would print this truncating log line twice which is not that clean, will edit